js/src/components/paid-ads/convertToAssetGroupUpdateBody.test.js

```diff
@@ -0,0 +1,450 @@
+
+			const body = convertToAssetGroupUpdateBody( assetGroup, values );
+
+			[
+				ASSET_FORM_KEY.BUSINESS_NAME,
+				ASSET_FORM_KEY.HEADLINE,
+				ASSET_FORM_KEY.LOGO,
+			].forEach( ( key ) => {
+				const operations = body.assets.filter(
+					( operation ) => operation.field_type === key
+				);
+				const asset = assetGroup.assets[ key ];
+				const value = values[ key ];
+
+				expect( operations ).toContainEqual( {
+					id: asset.id || asset[ 0 ].id,
+					content: null,
+					field_type: key,
+				} );
+				expect( operations ).toContainEqual( {
+					id: null,
+					content: Array.isArray( value ) ? value[ 0 ] : value,
+					field_type: key,
+				} );
+			} );
+		} );
+	} );
+} );
+
+describe( 'diffAssetOperations', () => {
+	const filledAssetGroup = {
+		[ ASSET_GROUP_KEY.FINAL_URL ]: 'https://example',
+		[ ASSET_GROUP_KEY.DISPLAY_URL_PATH ]: [ 'cap', 'specials' ],
+		assets: {
+			[ ASSET_KEY.BUSINESS_NAME ]: {
+				id: genId(),
+				content: 'General Clothing',
+			},
+			[ ASSET_KEY.MARKETING_IMAGE ]: [
+				{
+					id: genId(),
+					content: 'https://example/image-1.png',
+				},
+			],
+			[ ASSET_KEY.SQUARE_MARKETING_IMAGE ]: [
+				{
+					id: genId(),
+					content: 'https://example/square-image-1.png',
+				},
+			],
+			[ ASSET_KEY.PORTRAIT_MARKETING_IMAGE ]: [
+				{
+					id: genId(),
+					content: 'https://example/portrait-image-1.png',
+				},
+			],
+			[ ASSET_KEY.LOGO ]: [
+				{
+					id: genId(),
+					content: 'https://example/logo-1.png',
+				},
+				{
+					id: genId(),
+					content: 'https://example/logo-2.png',
+				},
+				{
+					id: genId(),
+					content: 'https://example/logo-3.png',
+				},
+			],
+			[ ASSET_KEY.HEADLINE ]: [
+				{
+					id: genId(),
+					content: 'headline 1',
+				},
+				{
+					id: genId(),
+					content: 'headline 2',
+				},
+				{
+					id: genId(),
+					content: 'headline 3',
+				},
+			],
+			[ ASSET_KEY.LONG_HEADLINE ]: [
+				{
+					id: genId(),
+					content: 'long headline 1',
+				},
+			],
+			[ ASSET_KEY.DESCRIPTION ]: [
+				{
+					id: genId(),
+					content: 'description 1',
+				},
+			],
+			[ ASSET_KEY.CALL_TO_ACTION_SELECTION ]: {
+				id: genId(),
+				content: 'sign_up',
+			},
+		},
+	};
+
+	const filledValues = {
+		[ ASSET_FORM_KEY.BUSINESS_NAME ]: 'General Clothing',
+		[ ASSET_FORM_KEY.MARKETING_IMAGE ]: [ 'https://example/image-1.png' ],
+		[ ASSET_FORM_KEY.SQUARE_MARKETING_IMAGE ]: [
+			'https://example/square-image-1.png',
+		],
+		[ ASSET_FORM_KEY.PORTRAIT_MARKETING_IMAGE ]: [
+			'https://example/portrait-image-1.png',
+		],
+		[ ASSET_FORM_KEY.LOGO ]: [
+			'https://example/logo-1.png',
+			'https://example/logo-2.png',
+			'https://example/logo-3.png',
+		],
+		[ ASSET_FORM_KEY.HEADLINE ]: [
+			'headline 1',
+			'headline 2',
+			'headline 3',
+		],
+		[ ASSET_FORM_KEY.LONG_HEADLINE ]: [ 'long headline 1' ],
+		[ ASSET_FORM_KEY.DESCRIPTION ]: [ 'description 1' ],
+		[ ASSET_FORM_KEY.CALL_TO_ACTION_SELECTION ]: 'sign_up',
+	};
```

Comment:
💅  Wonder if we could use this two as global CONSTANTS at the top. i.e. `DUMMY_ASSET_GROUP` and `DUMMY_FILLED_VALUES` shared between `convertToAssetGroupUpdateBody` and `diffAssetOperations` tests.

Reply:
They are slightly different, and the tests focus on different target functions, so changing to shared constants will cause the test code to require some extra processing for the sake of sharing, which will make the tests slightly out of focus.

However, I found some redundancies and an inconsistent default value when I revisited the code. Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/1874/commits/3e178c737de23fadf1702d99fbc7d99ac626b913.